AzureTable/Microsoft.DataTransfer.AzureTable/Sink/TableAPIBulkSinkAdapter.cs

```diff
@@ -0,0 +1,172 @@
+        private ConcurrentDictionary<string, TableBatchOperation> dict;
+        private InputSizeTracker inputSizeTracker;
+        private BatchSizeTracker batchSizeTracker;
+
+        public int MaxDegreeOfParallelism
+        {
+            get { return 1; }
+        }
+
+        public TableAPIBulkSinkAdapter(string connectionString, string tableName, 
+            bool overwrite, long maxInputBufferSizeInBytes, int throughput, int batchSize)
+        {
+            _connectionString = connectionString;
+            _tableName = tableName;
+            _overwrite = overwrite;
+            _maxInputBufferSizeInBytes = maxInputBufferSizeInBytes;
+            _throughput = throughput;
+            _maxLengthInBytesPerBatch = batchSize;
+
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(_connectionString);
+
+            TableConnectionPolicy connectionPolicy = new TableConnectionPolicy()
+            {
+                UseDirectMode = true,
+                UseTcpProtocol = true,
+            };
+
+            CloudTableClient tableClient = storageAccount.CreateCloudTableClient(connectionPolicy: connectionPolicy);
+            cloudtable = tableClient.GetTableReference(_tableName);
+        }
+
+        public async Task InitializeAsync(CancellationToken cancellation)
+        {
+            inputSizeTracker = new InputSizeTracker(maxLengthInBytesPerDocument, _maxInputBufferSizeInBytes);
+            batchSizeTracker = new BatchSizeTracker(_maxLengthInBytesPerBatch, inputSizeTracker);
+            dict = new ConcurrentDictionary<string, TableBatchOperation>();
+            await cloudtable.CreateIfNotExistsAsync(IndexingMode.Consistent, _throughput, cancellation);
+        }
+
+        public async Task WriteAsync(IDataItem dataItem, CancellationToken cancellation)
+        {
+            var item = GetITableEntityFromIDataItem(dataItem);
+
+            inputSizeTracker.Add(item);
+
+            if (!dict.ContainsKey(item.PartitionKey))
+            {
+                dict[item.PartitionKey] = new TableBatchOperation();
+            }
+
+            var batchOperation = dict[item.PartitionKey];
+            batchOperation.InsertOrReplace(item);
+
+            if (inputSizeTracker.HasExceededInputLimits())
+            {
+                await FlushToTableAsync(cancellation);
+            }
+        }
+
+        public async Task CompleteAsync(CancellationToken cancellation)
+        {
+            if (!inputSizeTracker.HasExceededInputLimits())
+            {
+                await FlushToTableAsync(cancellation);
+            }
+        }
+
+        public void Dispose()
+        {
+        }
+
+        // Flush out all data to the Cosmos DB Table sink.
+        private async Task FlushToTableAsync(CancellationToken cancellation)
+        {
+            try
+            {
+                List<Exception> exceptions = new List<Exception>();
+
+                foreach (var kv in dict)
```

Comment:
How about doing cross partition in parallel?

Reply:
As discussed over skype, this is not required for GA. 
We will integrate with the Data Migration library post GA, to address this.